[WIP] Feature mintty tui

erw7 opened the pull request:
This is an experimental implementation of mintty TUI which took over #7660.

~~After executeing, the state of the terminal does not return completely. Therefore drawing may be very slow. Please use `reset` or `stty` to restore the state of the terminal.~~

- [X] Resolved issue where `tcgetattr()` returns `struct __oldtermios` (This may be the cause of the terminal not returning completely to its original state).
- [x] Investigate the reason why `uv_write` fails to write to a pipe when `nbufs>=2`.
- [x] Confirm whether we can compile in environments other than MinGW (64 bit).
- [ ] Implementation of error handling.
- [ ] Confirm that there are no other problems.

Marwes committed:
Detect mintty as a terminal

Explanations

https://fossies.org/linux/vim/src/iscygpty.c
https://github.com/BurntSushi/ripgrep/issues/94#issuecomment-261745480

erw7 committed:
Implement mintty TUI

erw7 committed:
Fix issue that mintty TUI abnormally terminate

erw7 committed:
Corresponding to build at MSVC

justinmk commented on src/nvim/os/cygterm.c:
```diff
@@ -0,0 +1,302 @@
+// https://fossies.org/linux/vim/src/iscygpty.c
+// See https://github.com/BurntSushi/ripgrep/issues/94#issuecomment-261745480
+// for an explanation on why this works
+int query_mintty(int fd, MinttyQueryType query_type)
```
small note about style: functions in `os/*` should be prefixed with `os_`.

erw7 committed:
Change design

- Rename function.
- Add Cygwindll struct.
- Change function used only in cygterm.c to static function.
- Fix that the termination process was not called.
- Correct warning.
- Fix lint issue.
- Add comment.

erw7 committed:
Fix possibility that memory corruption

erw7 committed:
Add code that update window size

erw7 committed:
Change to enable termination processing of terminal

janlazo commented:
Does `bash -c nvim` work now so that nvim can run old-style tests on sh/bash?

justinmk commented:
> This assumes that TERM is unset so that sh/bash can set it to TERM=cygwin.

Is there a terminfo file for "cygwin" ?

janlazo commented:
I have `/usr/share/terminfo/63/cygwin` for msysgit on my machine. It's the only file in that folder while the other folders (64/, 78/) have `dumb`, or variants of `xterm`.

erw7 committed:
Fix condition for initializing dll of cygwin

erw7 committed:
Move some definitions from header to source

erw7 committed:
Change to set API version of Cygwin

The cause of tcgetattr() returning __oldtermios was that API version of
Cygwin was not set.

erw7 committed:
Fix lint issue

erw7 committed:
Remove unnecessary comment

erw7 committed:
Fix issue that terminate abnormally with SIGSEGV

erw7 committed:
Organize code

- Remove unnecessary definition.
- Change not to expose to externally the inside of sturct CygTerm.
- Move some definition to source file from header.

erw7 committed:
Remove unnecessary include file

justinmk commented on src/nvim/os/cygterm.h:
```diff
@@ -0,0 +1,21 @@
+  kNoneMintty,
+  kMinttyCygwin,
+  kMinttyMsys
+} MinttyType;
```
why `kNoneMintty` instead of `kMinttyNone` ?

erw7 committed:
Fix enumeration constant

erw7 committed:
Fix issue with size of utsname

erw7 committed:
Edit comment

erw7 committed:
Fix issue that padding is not allocate with MSVC

erw7 committed:
Fix issue that build failed with MSVC due to locale

erw7 committed:
Fix order of function

erw7 commented:
@janlazo  This PR is to work nvim on Mintty.  I think that you would like to run the old-style test on the console(cmd.exe).  Therefore, I think this PR is irrelevant. `bash -c nvim` will work on the console even without this PR.

janlazo commented:
@erw7 The new-style oldtests now use `set shell=cmd.exe` and ignore the value of `$SHELL` because nvim segfaults on sh/bash (ie `bash -c nvim` segfaults). I assume sh/bash depends on msys2/cygwin environment to work and mintty uses sh/bash by default such that this PR will allow nvim to use `set shell=sh` in `:terminal` on Windows. Use case is running the old-style oldtests with minimal changes to the `src/nvim/testdirMakefile`.

erw7 commented:
@janlazo I remembered that I fixed the problem with [this commit](https://github.com/neovim/neovim/pull/8271/commits/dc17b869795f8954f994f03abe87cf6d285157a3) that SIGSEGV occurs when `TERM` is set in the neovim bundled unibilium. It occurred on [this line](https://github.com/mauke/unibilium/blob/v2.0.0/uniutil.c#L200) of `unibi_from_term()`, because TERMINFO_DIRS was not defined. It does not happen if `TERMINFO` is properly set, so it may be better to set it to `TERMINFO=/usr/share/terminfo`.

erw7 committed:
Change design

- Change to exit when loading cygwin dll fails.
- Hide CygwinDll detail from CygTerm.
- Unify the prefix of the function name.
- Add comment.
- Change not to call functions in cygwin dll from other than initialized
  thread.

janlazo commented:
@erw7 When does nvim check `TERMINFO`? Can the user set on startup via init.vim so that the user doesn't have to use a shell script or batchfile to set environment variables prior to running nvim?

I'm thinking that it's possible to workaround Windows TUI problems by requiring sh/bash and the terminfo for `TERM=cygwin` to properly render the screen for end-users since there's no terminfo for Windows's default terminal.

justinmk commented:
> Can the user set on startup via init.vim so that the user doesn't have to use a shell script or batchfile to set environment variables prior to running nvim?

Check where ui_builtin_start is called from main.c .

justinmk commented:
It happens after source_startup_scripts so init.vim should work to set TERMINFO.

janlazo commented:
Can the patch to prevent abnormal termination because of `TERM` be cherry-picked for 0.3?

justinmk commented:
sure, can someone cherry-pick it to a new PR?

erw7 commented:
I have created a [new pull request](https://github.com/neovim/neovim/pull/8408) that only fix issue with TERM.

@janlazo I think TUI display on WIndows 7 will improve, by adding `cygwin` to the built-in TERMINFO and using it. But I think there is a possibility that `xtem-256colors` is better on Windows 10.

janlazo commented:
nvim TUI is usable on bash with the `TERM` fix. Scrolling is slow with non-default colorscheme. Which terminfo is `nvim -u NORC` using with `TERM=cygwin`?

justinmk commented:
@janlazo try infocmp, it should show the terminfo file at the top of its output:

    infocmp -L

janlazo commented:
```
#	Reconstructed via infocmp from file: /usr/share/terminfo/63/cygwin
cygwin|ansi emulation for Cygwin,
	am, hs, mir, msgr, xon,
	colors#8, it#8, pairs#64,
	acsc=+\020\,\021-\030.^Y0\333`\004a\261f\370g\361h\260j\331k\277l\332m\300n\305o~p\304q\304r\304s_t\303u\264v\301w\302x\263y\363z\362{\343|\330}\234~\376,
	bel=^G, bold=\E[1m, clear=\E[H\E[J, cr=\r, cub=\E[%p1%dD,
	cub1=^H, cud=\E[%p1%dB, cud1=\E[B, cuf=\E[%p1%dC,
	cuf1=\E[C, cup=\E[%i%p1%d;%p2%dH, cuu=\E[%p1%dA,
	cuu1=\E[A, dch=\E[%p1%dP, dch1=\E[P, dl=\E[%p1%dM,
	dl1=\E[M, ed=\E[J, el=\E[K, el1=\E[1K, fsl=^G, home=\E[H,
	hpa=\E[%i%p1%dG, ht=^I, ich=\E[%p1%d@, ich1=\E[@,
	il=\E[%p1%dL, il1=\E[L, ind=\n, invis=\E[8m, kb2=\E[G,
	kbs=^H, kcub1=\E[D, kcud1=\E[B, kcuf1=\E[C, kcuu1=\E[A,
	kdch1=\E[3~, kend=\E[4~, kf1=\E[[A, kf10=\E[21~,
	kf11=\E[23~, kf12=\E[24~, kf13=\E[25~, kf14=\E[26~,
	kf15=\E[28~, kf16=\E[29~, kf17=\E[31~, kf18=\E[32~,
	kf19=\E[33~, kf2=\E[[B, kf20=\E[34~, kf3=\E[[C, kf4=\E[[D,
	kf5=\E[[E, kf6=\E[17~, kf7=\E[18~, kf8=\E[19~, kf9=\E[20~,
	khome=\E[1~, kich1=\E[2~, knp=\E[6~, kpp=\E[5~, kspd=^Z,
	nel=\r\n, op=\E[39;49m, rc=\E8, rev=\E[7m, ri=\EM,
	rmacs=\E[10m, rmcup=\E[2J\E[?47l\E8, rmir=\E[4l,
	rmpch=\E[10m, rmso=\E[27m, rmul=\E[24m, rs1=\Ec\E]R,
	sc=\E7, setab=\E[4%p1%dm, setaf=\E[3%p1%dm,
	sgr=\E[0;10%?%p1%t;7%;%?%p2%t;4%;%?%p3%t;7%;%?%p6%t;1%;%?%p7%t;8%;%?%p9%t;11%;m,
	sgr0=\E[0;10m, smacs=\E[11m, smcup=\E7\E[?47h,
	smir=\E[4h, smpch=\E[11m, smso=\E[7m, smul=\E[4m, tsl=\E];,
	u6=\E[%i%d;%dR, u7=\E[6n, u8=\E[?6c, u9=\E[c,
	vpa=\E[%i%p1%dd,
```

justinmk commented:
@janlazo There's no builtin for cygwin (yet), see terminfo_builtin: 
 https://github.com/neovim/neovim/blob/60716371e97dae916c6525e9ba840aae562069bb/src/nvim/tui/terminfo.c#L105

erw7 commented:
@justinmk You used to say [this](https://github.com/neovim/neovim/issues/8315#issuecomment-384039087) before.

@janlazo neovim looks for terminfo in [`unibi_from_term`](https://github.com/mauke/unibilium/blob/master/uniutil.c#L174) as follows.

- `$TERMINFO`
- `${HOME}/.terminfo`
- `$TERMINFO_DIRS`
- `unibi_terminfo_dirs` (Empty on Windows)

If terminfo is not found anywhere above, use builtin terminfo. In that case, since the builtin terminfo does not have cygwin, ansi_terminfo is used.

janlazo commented:
`$TERMINFO`, `$TERMINFO_DIRS` were not set and `${HOME}/.terminfo` doesn't exist so it's using ansi_terminfo on my machine? I tried setting `$TERMINFO` to Git's `/usr/share/terminfo` but I can't tell the difference.

janlazo commented:
@justinmk See https://gist.github.com/janlazo/975f580d84c5f6446c09e4dd694c0d47

janlazo commented:
ConEmu + bash is required for the TUI atm. Using bash on default Windows terminal doesn't display the cursor. See https://github.com/janlazo/dotvim8/commit/1e2eea81f3934c49c513b9623e15b074ef88be2b

justinmk commented:
> @justinmk See https://gist.github.com/janlazo/975f580d84c5f6446c09e4dd694c0d47

Ok, that confirms that Nvim isn't loading `cygwin` terminfo for some reason:

```
&term: builtin_ansi
Description: ansi/pc-term compatible with color
```

I wonder why `unibi_from_env()` doesn't find it.

erw7 commented:
Looking at this [log](https://gist.github.com/janlazo/975f580d84c5f6446c09e4dd694c0d47#file-log-cygwin-txt-L1214), `cygwin` terminfo seems to be found.

@janlazo If you add `cnorm` etc. in the following procedure, will the cursor be displayed?

```
infocmp cygwin > cygwin.ti
```

Added `cnorm=\E[?25h,`, `civis=\E[?25l,`,  `cvvis=\E[?25h,` to cygwin.ti with a editor.

```
tic -o ${HOME}/.terminfo cygwin.ti
```

unset `$TERMINFO`.

janlazo commented:
I have Windows 8.1 so I can't use `TERM=xterm`.

Before adding it, the screen was normal with vertical cursor (ConEmu) or block cursor (cmd.exe). I can't highlight on ConEmu for some reason.
After adding it and creating `~/.terminfo`, Vim's and Neovim's cursor broke, outputting escape sequences instead.

erw7 commented:
> After adding it and creating ~/.terminfo, Vim's and Neovim's cursor broke, outputting escape sequences instead.

I think you have failed to edit terminfo.

I would like to sort out a problem. Is your problem as follows?

1. Scrolling is slow with non-default colorscheme.
2. Using bash on default Windows terminal doesn't display the cursor. 
3. On powershell with unset TERM, the TUI is still unusable.
4. I can't highlight on ConEmu for some reason.

janlazo commented:
Yes.

erw7 commented:
Information is too fragmented to reproduce everything in my environment.

Are these problems caused by nvim applying this PR? Or are those problems going to happen in mintty? Otherwise, I think here is not very appropriate for discussion. Would you be able to create a new issue if possible?

justinmk commented:
@erw7 We could merge this without finishing the TODO list, if it improve things without breaking anything. Would be nice to make the changes to `tui/tui.c` and `ui_bridge.c` somehow more isolated, if possible.

erw7 commented:
@justinmk I'm sorry I can't get time, and my reply is delayed. I am skeptical about merging this PR after reading https://sourceware.org/ml/cygwin/2014-11/msg00411.html. Also, I think that the function for calling `cygwin1.dll` from a Windows native program is not maintained. So I do not intend to continue this work. If there is a person who needs this function, I would like that person to take over this work.
